src/algebra/star/chsh.lean

```diff
@@ -0,0 +1,274 @@
+      repeat { conv in (B₀ * (A₁ * B₀))
+      { rw [T.A₁B₀_commutes, ←mul_assoc B₀ B₀ A₁, ←pow_two, T.B₀_inv, one_mul], } },
+      repeat { conv in (B₁ * (A₀ * B₁))
+      { rw [T.A₀B₁_commutes, ←mul_assoc B₁ B₁ A₀, ←pow_two, T.B₁_inv, one_mul], } },
+      repeat { conv in (B₁ * (A₁ * B₁))
+      { rw [T.A₁B₁_commutes, ←mul_assoc B₁ B₁ A₁, ←pow_two, T.B₁_inv, one_mul], } },
+      conv in (A₀ * (B₀ * (A₀ * B₁)))
+      { rw [←mul_assoc, T.A₀B₀_commutes, mul_assoc, ←mul_assoc A₀, ←pow_two, T.A₀_inv, one_mul], },
+      conv in (A₀ * (B₁ * (A₀ * B₀)))
+      { rw [←mul_assoc, T.A₀B₁_commutes, mul_assoc, ←mul_assoc A₀, ←pow_two, T.A₀_inv, one_mul], },
+      conv in (A₁ * (B₀ * (A₁ * B₁)))
+      { rw [←mul_assoc, T.A₁B₀_commutes, mul_assoc, ←mul_assoc A₁, ←pow_two, T.A₁_inv, one_mul], },
+      conv in (A₁ * (B₁ * (A₁ * B₀)))
+      { rw [←mul_assoc, T.A₁B₁_commutes, mul_assoc, ←mul_assoc A₁, ←pow_two, T.A₁_inv, one_mul], },
+      simp only [←pow_two, T.A₀_inv, T.A₁_inv],
+      simp only [mul_comm A₁ A₀, mul_comm B₁ B₀, mul_left_comm A₁ A₀, mul_left_comm B₁ B₀,
+        mul_left_comm B₀ A₀, mul_left_comm B₀ A₁, mul_left_comm B₁ A₀, mul_left_comm B₁ A₁],
+      norm_num,
+      simp only [mul_comm _ (2 : R), mul_comm _ (4 : R),
+        mul_left_comm _ (2 : R), mul_left_comm _ (4 : R)],
+      abel,
+      simp only [neg_mul_eq_neg_mul_symm, mul_one, int.cast_bit0, one_mul, int.cast_one,
+        gsmul_eq_mul, int.cast_neg],
+      simp only [←mul_assoc, ←add_assoc],
+      norm_num, },
+    have idem' : P = (1 / 4 : ℝ) • (P * P),
+    { have h : 4 * P = (4 : ℝ) • P := by simp [algebra.smul_def],
+      rw [idem, h, ←mul_smul],
+      norm_num, },
+    have sa : star P = P,
+    { dsimp [P],
+      simp only [star_add, star_sub, star_mul, star_bit0, star_one,
+        T.A₀_sa, T.A₁_sa, T.B₀_sa, T.B₁_sa, mul_comm B₀, mul_comm B₁], },
+    rw idem',
+    conv_rhs { congr, skip, congr, rw ←sa, },
+    convert smul_le_smul_of_nonneg (star_mul_self_nonneg : 0 ≤ star P * P) _,
+    { simp, },
+    { apply_instance, },
+    { norm_num, }, },
+  apply le_of_sub_nonneg,
+  simpa only [sub_add_eq_sub_sub, ←sub_add] using i₁,
+end
+
+/-!
+Before proving Tsirelson's bound,
+we prepare some easy lemmas about √2.
+-/
+
+
+local notation `√2` := (2^(2⁻¹ : ℝ) : ℝ)
+
+lemma one_lt_sqrt_two : 1 < √2 :=
+by { convert real.rpow_lt_rpow _ one_lt_two _; norm_num }
+
+lemma sqrt_two_ne_zero : √2 ≠ 0 :=
+ne_of_gt (lt_trans zero_lt_one one_lt_sqrt_two)
+
+lemma zero_lt_sqrt_two_inv : 0 < √2⁻¹ :=
+begin
+  simp only [inv_pos],
+  apply real.rpow_pos_of_pos,
+  norm_num,
+end
+
+@[simp] lemma sqrt_two_inv_sq : √2⁻¹ * √2⁻¹ = (2⁻¹ : ℝ) :=
+by { rw [←mul_inv', ←real.rpow_add]; norm_num }
+
+/-!
+We next need some silly lemmas about numerals in modules and algebras.
+If anyone sees how to obtain these from general statements, please improve this!
+-/
+
+lemma two_gsmul_half_smul {α : Type*} [add_comm_group α] [module ℝ α] {X : α} :
+  2 •ℤ (2⁻¹ : ℝ) • X = X :=
```

Comment:
I guess with our improvements in `is_scalar_tower` and friends, you can now use regular `smul` here, instead of `gsmul`.

Reply:
I'm not sure what you have in mind here. Changing the statement wouldn't help --- later on I'm faced with these grungy expressions involving both versions of scalar multiplication, because `abel` produces them.

Reply:
I see, that makes sense. I would still prefer to have these namespaced. Especially if they are so specialized about `real`.

Reply:
I've hidden them in a namespace, and added more comments.